exporter/splunkhecexporter/client.go

```diff
@@ -389,22 +389,33 @@ func (c *client) pushMetricsRecords(ctx context.Context, mds pmetric.ResourceMet
 	metrics := res.ScopeMetrics().At(state.library).Metrics()
 	bufCap := int(c.config.MaxContentLengthMetrics)
 
+	var events []*splunk.Event
 	for k := 0; k < metrics.Len(); k++ {
+		// Parsing metric record to Splunk event.
+		events = append(events, mapMetricToSplunkEvent(res.Resource(), metrics.At(k), c.config, c.logger)...)
+	}
+
+	if c.config.UseMultiMetricFormat {
+		merged, err := mergeEventsToMultiMetricFormat(events)
+		if err != nil {
+			permanentErrors = append(permanentErrors, consumererror.NewPermanent(fmt.Errorf("error merging events: %w", err)))
+		} else {
+			events = merged
+		}
+	}
+
+	for k := 0; k < len(events); k++ {
 		if state.bufFront == nil {
 			state.bufFront = &index{resource: state.resource, library: state.library, record: k}
```

Comment:
Is this ok that we store event index in `state.bufFront.record` instead of metric index now? Asking just in case, not sure if any functionality can be affected. The values must be different.

Reply:
this code is too clever. Looking through it one more time.

Reply:
Yes, it's impacting. The code used to guarantee reporting the subset of metrics that could not be sent by keeping track of the last index of metrics sent. With this change, it can no longer guarantee that, and it's hard to keep track of which metrics make which events.

I need to think on this a bit more.

Reply:
@dmitryax @atoulme I think we should segregate preparing batches and sending events. Attaching a high-level diagram for log data.

Current implementation:
![image](https://user-images.githubusercontent.com/89519921/200829698-9b1d8cad-ca1c-46d8-bad6-0208afa1be7b.png)

Suggested implementation:
![image](https://user-images.githubusercontent.com/89519921/200829842-445e4775-3ac1-4e81-99a3-d5fd4d07f4f5.png)

The suggested approach has some advantages, like:
- It will not have the overhead recreating sub log or sub metric in case of failure
- All events will be marshalled and compressed only once
- The design fits well with the HEC indexer acknowledgement

Reply:
I wanted to get https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/15383 in first before revisiting, as it was more impactful and might change the way data is handled. I would defer to you on the best approach here.